File: movieui/src/scripts/ws/Film/getFilm.ts
import {wsUrlRoot} from "@/scripts/ws/meta"
import {createWebSocket, recvMsg, reqStringify, rspParse, sendMsg} from "@/scripts/ws/helper"

export {getFilm}
export type {
    GetFilmReq,
    GetFilmRsp
}

type GetFilmReq =
    {
        FilmId: bigint
    }

type GetFilmRsp =
    {
        Ok: boolean
        FilmName: string
        FilmCoverUrl: string
        FilmPreviewVideoUrl: string
        FilmPrice: number
        FilmIsPreorder: boolean
        FilmOnlineTime: Date
    }

async function getFilm(req: GetFilmReq) {
    const conn = createWebSocket(`${wsUrlRoot}/get_film`)

    const task = recvMsg(conn)
    const reqJson = reqStringify(req)
    console.log('get_film req:' + reqJson)
    sendMsg(conn, reqJson).then()
    const msg = await task
    console.log('get_film rsp:' + msg)

    return <GetFilmRsp>rspParse(msg)
}

File: SuperMovie_server/src/api/Film/GetFilm.cs
namespace SuperMovie.Server.Api.Film;

using Container.Film.Provider;
using WebSocketSharp;
using WebSocketSharp.Server;
using Util;

public struct GetFilmReq
{
    public long FilmId;
}

public struct GetFilmRsp
{
    public bool Ok;
    public string FilmName;
    public string FilmCoverUrl;
    public string FilmPreviewVideoUrl;
    public double FilmPrice;
    public bool FilmIsPreorder;
    public DateTime FilmOnlineTime;
    public List<string> FilmTypes;
    public List<string> FilmActors;
}

//api : get_film
public class GetFilm : WebSocketBehavior
{
    private IFilmProvider _filmProvider;

    public void Set(IFilmProvider filmProvider)
    {
        _filmProvider = filmProvider;
    }

    protected override void OnMessage(MessageEventArgs e)
    {
        Console.WriteLine($"get_film req:\n{e.Data}");

        var req = JsonHelper.Parse<GetFilmReq>(e.Data);
        var film = _filmProvider.GetFilm(req.FilmId);

        GetFilmRsp rsp;

        if (film != null)
        {
            rsp = new GetFilmRsp
            {
                Ok = true,
                FilmName = film.Name,
                FilmCoverUrl = film.CoverUrl,
                FilmPreviewVideoUrl = film.PreviewVideoUrl,
                FilmPrice = film.Price,
                FilmIsPreorder = film.IsPreorder,
                FilmOnlineTime = film.OnlineTime.Value,
                FilmTypes = film.Types,
                FilmActors = film.Actors
            };
        }
        else
        {
            rsp = new GetFilmRsp
            {
                Ok = false,
                FilmName = "",
                FilmCoverUrl = "",
                FilmPreviewVideoUrl = "",
                FilmPrice = -1,
                FilmIsPreorder = false
            };
        }

        var json = JsonHelper.Stringify(rsp);
        Console.WriteLine($"get_film rsp:\n{json}");
        Send(json);
    }
}